base/array.jl

```diff
@@ -232,6 +232,11 @@ collect(itr) = collect(eltype(itr), itr)
 ## Iteration ##
 start(A::Array) = 1
 next(a::Array,i) = (a[i],i+1)
+function unsafe_next(a::Array,i)
+    @inbounds begin
+        return next(a,i)
```

Comment:
Not sure if this is still the case, but `@inbounds return` was difficult for inference to reason about at one point in recent history.  You're probably more aware of the current design than I am, but just FYI: https://github.com/JuliaLang/julia/pull/13461.